Testing pipeline

JosePizarro3 opened the pull request:
I am testing the pipeline and restructuring a bit the project to look more aligned with the central NOMAD and other plugins which are integrated in the software.

JosePizarro3 committed:
Restructure src folder into analysis

Deleted setup.py

Deleted MANIFEST

Change naming in yaml config files

Merge pyproject toml with NOMAD installation

JosePizarro3 committed:
Skipping test

JosePizarro3 committed:
Fixing pipeline

JosePizarro3 committed:
Fix pyproject.toml

JosePizarro3 committed:
Updated README

JosePizarro3 committed:
Added NOMAD badge

JosePizarro3 committed:
Trying naming everything analysis

Fixing ruff

JosePizarro3 committed:
Moved formatting BEFORE check

JosePizarro3 committed:
fix missing *

JosePizarro3 committed:
Try 3

JosePizarro3 commented:
@ladinesa @ka-sarthak any idea why the pipeline does not work? I just renamed `src` to `analysis` and moved the modules one level up (before they were under `src/nomad_analysis`. 

Everything else, I tried to copy from the example, so I am really puzzled.

Btw, as an additional question: do we need to test parsing from nomad.cli? I think we can get rid off these in this repo anyways.

ka-sarthak commented:
Hi @JosePizarro3

I had similar test failures earlier and it was probably because the nomad parse was unable to find the analysis schema. If you export the path of the new src (or `analysis` in this case) to PYTHONPATH, parse function might work. A better solution is to add the name of folder containing the package code to this quantity in `pyproject.toml` and then re-build the package.
```toml
[tool.setuptools.packages.find]
where = ["src"]
```
But the choice for having `src/nomad-analysis` was a deliberate one. It's a good practice to have src folder that contains the package code (read more [here](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)). I also raised an issue to support this in `nomad-schema-plugin-example` of which this repo is a fork. https://github.com/nomad-coe/nomad-schema-plugin-example/issues/9 

Similar structure is being used for `nomad-measurements`, `nomad-material-properties`, and other plugins being developed in Area A.

JosePizarro3 commented:
> Hi @JosePizarro3
> 
> I had similar test failures earlier and it was probably because the nomad parse was unable to find the analysis schema. If you export the path of the new src (or `analysis` in this case) to PYTHONPATH, parse function might work. A better solution is to add the name of folder containing the package code to this quantity in `pyproject.toml` and then re-build the package.
> 
> ```toml
> [tool.setuptools.packages.find]
> where = ["src"]
> ```
> 
> But the choice for having `src/nomad-analysis` was a deliberate one. It's a good practice to have src folder that contains the package code (read more [here](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)). I also raised an issue to support this in `nomad-schema-plugin-example` of which this repo is a fork. [nomad-coe#9](https://github.com/nomad-coe/nomad-schema-plugin-example/issues/9)
> 
> Similar structure is being used for `nomad-measurements`, `nomad-material-properties`, and other plugins being developed in Area A.

Yeah, thanks for the info. My question is why renaming does not work, and the plugin example works with a flat folder structure (and all the other plugins we have ...).

Personally I am not friend of overnesting stuff (we are essentially having 2 folders instead of a flat hierarchy. But I am happy with whatever. We just need to be consistent.

JosePizarro3 commented:
One remark on the structure: I think `nomad-analysis` could be itself a placeholder for individual plugins which then get hooked up; something like:
```
.
└── nomad-analysis/
    └── src/
        ├── spectral_profiles_analysis/
        │   ├── nomad_plugin.yaml
        │   └── ...
        ├── descriptors_analysis/
        │   ├── nomad_plugin.yaml
        │   └── ...
        └── <whatever>_analysis/
            ├── nomad_plugin.yaml
            └── ...
```

Still, I am puzzled by the fact that the pipeline mimicking the nomad-parser-plugin-example does not work. I will keep investigating tomorrow.

ka-sarthak commented on README.md:
```diff
@@ -1,27 +1,32 @@
+git clone https://github.com/FAIRmat-NFDI/nomad-analysis.git
+cd nomad-analysis
 python3.9 -m venv .pyenv
-source .pyenv/bin/activate
```
Is there a reason to omit this?

JosePizarro3 replied:
I forgot 😄 thanks for pointing it out

ka-sarthak commented on README.md:
````diff
@@ -32,35 +37,34 @@ You can run automated tests with `pytest`:
-```sh
-pip install -e .[dev] --index-url https://gitlab.mpcdf.mpg.de/api/v4/projects/2187/packages/pypi/simple
-```
 
````
I think ideally we should have some instructions pointing to using the package in `editable` mode when developing. It will be important when having new developers onboard.

JosePizarro3 replied:
What do you mean? I thought these lines were pretty much repetition of the "Install the dependencies" part.

ka-sarthak replied:
There is a flag `-e` that differs, which installs the package in editable mode. So that when you make some changes to the code in the package, it reflects in the installed package as well.
https://setuptools.pypa.io/en/latest/userguide/development_mode.html

JosePizarro3 replied:
Ah true, you are right. My bad, I will restore the Development section.

ka-sarthak commented on setup.py:
This file is needed to install the package on IKZ oasis.

JosePizarro3 replied:
Yeah, indeed, I was getting crazy to see whether deleting or not will change something on the pipeline. This is to be restored.

ka-sarthak commented on MANIFEST.in:
This file is needed to install the package on IKZ oasis.

JosePizarro3 replied:
Is it tho? I think the MANIFEST only calls for specific plugins to be installed, hence it should be in NOMAD, right? How does it exactly work in OASIS?

ka-sarthak replied:
I am not exactly sure why, but this fixed some issues while adding analysis plugin to the oasis. @hampusnasstrom might know more about this

hampusnasstrom replied:
The MANIFEST.in is just a file telling setuptools what files (in addition to .py files) that should be included when packaging up a python package. Since the `nomad_plugin.yaml` is not a python file and not imported by any of the python files in the plugin it might be ignored without a MANIFEST.in specifically including it.

JosePizarro3 replied:
Let me say this way: why there are not any other MANIFEST.in files in the other plugin repos? I am guessing that this file can be modified by the OASIS admin. Is this correct? If so, I'll delete this file and simply give instructions on how to add `nomad-analysis` to the corresponding MANIFEST.in

hampusnasstrom replied:
Because they are currently included as submodules but all plugins will eventually be pip installed as python packages.
The MANIFEST.in has nothing to do with whether the plugin is activated or not. All that is handled by the nomad.yaml. The MANIFEST.in is just making sure that setuptools knows what to package up when pip installing this python package. Please just leave it there!

ka-sarthak commented on tests/test_schema.py:
```diff
@@ -1,11 +1,31 @@
 from nomad.client import normalize_all, parse
 
 
+# @pytest.mark.skip(reason="These tests should only be for the schema validation and functionalities, not the actual data normalization and parsing")
```
I agree there should be more tests for schema validation and functionalities as you said. This test can be a placeholder until we add more useful tests.

JosePizarro3 replied:
Indeed 🙂

ka-sarthak commented:
Hi @JosePizarro3, I tested the branch in my local and the pytest runs successfully. Even the one using coverage which seems to be failing the action in the workflow. 
```sh
python -m coverage run -m pytest -sv
```
I also added some comments on regarding the changes you made. Please review them :) 

Regarding the file structure rearrangement, I will still argue in favor of the src-layout, as it makes the testing of the *installed* package more robust. Also, I think we should have a `nomad_analysis` module within `src` to allows for much nicer import statement.
```sh
.
└── src/
    └── nomad_analysis/
        ├── spectral_profiles/
        │   ├── nomad_plugin.yaml
        │   └── ...
        ├── descriptors/
        │   ├── nomad_plugin.yaml
        │   └── ...
        └── <whatever>/
            ├── nomad_plugin.yaml
            └── ...
```
As the installed package will use everything inside /src, the import statements for the installed package would look like:
```py
from nomad_analysis.spectral_profiles import some_class
```
I think this is more descriptive and navigable than a generic `analysis`.

andreaa93 commented:
@JosePizarro3 don't screw up the plugin structure we have here, we can have a meeting and dicsuss why it is like that before you change it

JosePizarro3 commented:
> @JosePizarro3 don't screw up the plugin structure we have here, we can have a meeting and dicsuss why it is like that before you change it

Hey, this is just a draft where I didn't want any review yet, nor comments except for my questions.

I wanted to play with the pipeline, because the parser-plugin-example has a very different layout. I want to learn and standarize the way of developing and distributing plugins, so like I said to @ka-sarthak (in private), I am totally ok for using this layout, and adapt this to any other plugin.

andreaa93 commented:
> > @JosePizarro3 don't screw up the plugin structure we have here, we can have a meeting and dicsuss why it is like that before you change it
> 
> Hey, this is just a draft where I didn't want any review yet, nor comments except for my questions.
> 
> I wanted to play with the pipeline, because the parser-plugin-example has a very different layout. I want to learn and standarize the way of developing and distributing plugins, so like I said to @ka-sarthak (in private), I am totally ok for using this layout, and adapt this to any other plugin.

Oki 😇 Just thought you wanted to change it quick. Anyway, as Sarthak may have explained already, it's mainly for properly pip installability of the plugin, I also didn't enjoy much nesting, but then I found is more convenient when using it

Looking forward to see how this plugin evolves

JosePizarro3 commented:
> Regarding the file structure rearrangement, I will still argue in favor of the src-layout, as it makes the testing of the _installed_ package more robust. Also, I think we should have a `nomad_analysis` module within `src` to allows for much nicer import statement.
> 
> ```shell
> .
> └── src/
>     └── nomad_analysis/
>         ├── spectral_profiles/
>         │   ├── nomad_plugin.yaml
>         │   └── ...
>         ├── descriptors/
>         │   ├── nomad_plugin.yaml
>         │   └── ...
>         └── <whatever>/
>             ├── nomad_plugin.yaml
>             └── ...
> ```
> 
> As the installed package will use everything inside /src, the import statements for the installed package would look like:
> 
> ```python
> from nomad_analysis.spectral_profiles import some_class
> ```
> 
> I think this is more descriptive and navigable than a generic `analysis`.

Yeah, let's keep it src/nomad_analysis. As for now we only have spectra analysis, we keep it without further folder structure, and later we can move modules around.

Ok, I will restore the organization, setup.py, and just change a bit the code. Then it will be ready for review 😛

JosePizarro3 committed:
Added check in Ruff formatting

JosePizarro3 committed:
Improved README

JosePizarro3 committed:
Fixed repo url in mkdocs

hampusnasstrom commented on pyproject.toml:
```diff
@@ -1,65 +1,57 @@
 ]
 license = { text = "Apache-2.0" }
 requires-python = ">=3.9"
-classifiers = [
-    "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3 :: Only",
-    "Programming Language :: Python :: 3.9",
-    "License :: OSI Approved :: Apache Software License",
-]
```
What is the reason for removing the classifiers?

JosePizarro3 replied:
It surprises me we have to add these manually... and might make maintenance of the packages very cumbersome. I would assume that these are defined in other parts of the TOML, or better said, that pypi can grep these from somewhere.

I will read more next week.

JosePizarro3 replied:
Ok, I understood that these are automatically assigned:

https://python-poetry.org/docs/pyproject/#classifiers

https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#classifiers

However, I am not fully against defining them (and I can restore them back if you prefer so), thought I reiterate my concern: that maintaining these for ~60-70 plugins might be, in practice, very annoying.

hampusnasstrom commented on pyproject.toml:
```diff
@@ -1,65 +1,57 @@
 
-[project.urls]
-"Homepage" = "https://github.com/FAIRmat-NFDI/nomad-analysis"
-"Bug Tracker" = "https://github.com/FAIRmat-NFDI/nomad-analysis/issues"
```
What is the reason for excluding the bug tracker?

JosePizarro3 replied:
We can add as much as we want. Also I would suggest also to add "Documentation"

JosePizarro3 replied:
Ok, I added them back, and also "Documentation". Check it out in lines 24-29

hampusnasstrom commented on tests/test_schema.py:
```diff
@@ -1,11 +1,33 @@
 from nomad.client import normalize_all, parse
 
 
+@pytest.mark.skip(
+    reason='These tests should only be for the schema validation and functionalities, not the actual data normalization and parsing'
+)
```
I don't understand this. The normalization is here containing a lot of the functionality of the `ELNJupyterAnalysis`. Don't we need to test this @ka-sarthak?

JosePizarro3 replied:
No, this is not the case.

Testing here should test functionalities of the classes defined in src. For now, it is only parsing, and I would say that yes, we need to extend this.

For now, I would skip it tho. It is not testing anything.

hampusnasstrom replied:
I completely disagree. This might be how it was traditionally done but we are having a lot of *plugin functionallity* in the normalize function, such as creating the `.ipynb` file. This definitely needs to be tested to ensure that the plugin is working.

JosePizarro3 replied:
This is not what the test is testing....

I don't refer to what we traditionally did or not. But if we want to do testing, ok for me, but this is not testing what you are refering to.

hampusnasstrom replied:
We are testing that the normalization runs on line 31 `normalize_all(entry_archive)`... I agree that we should add more specific tests that the normalization actually yielded the results we expected. Also we need to add the `no_warn` fixture. @ka-sarthak please fill Chema in on what we discussed.

JosePizarro3 replied:
This is not testing `normalize()` from the `ELNJupyterAnalysis` class. `normalize_all()` calls for the plugins defined in nomad/config, and as this one is a standalone package, it is not included there and hence no normalization is ran.

The test is just checking that `EntryArchive.data.analysis_type == 'test'`, which I understood was a simple dummy test added for later extend this.

Hence, I again say: **this is not testing anything from the plugin**.

ka-sarthak replied:
For the ELN classes (which are part of the `src/nomad_analysis/schema.py`) that generate the JupyterNotebooks and the metadata around it, we will need to test the normalization. To do this test `nomad.normalize_all` is needed (will be further extended to include more sensible tests). Testing the parser is, however, optional as I understand because we aren't really writing parsers. But using `nomad.parse` simplifies the creation of input needed for `nomad.normalize_all`. I was a bit unsure about this machinery before, sorry for that @JosePizarro3. But we will need `test_schema` to be part of the testing.

Something I almost forgot about, which @hampusnasstrom just reminded, is that `nomad.normalize_all` will not throw any errors to the `test_schema` even if something is wrong. This is because it handles the errors by adding them to the logger. As such the tests will always pass. 

We need to add a `no_warn` fixture that digs up for error statements in the logs and raise Assertions if errors are found.

JosePizarro3 replied:
Again, you are not testing any normalization as of now because it will never run. The testing only constrains to checking that the key defined in `test.archive.yaml` is properly parsed.

In order to properly test the normalization, you must first import `ELNJupyterAnalysis` and, then, substitute `normalize_all(entry_archive)` by `ELNJupyterAnalysis.normalize(entry_archive)`. BUT, I doubt that this will run because:

1. You need to set the logger to None as default in each of your functions OR define it in your testing modules.
2. `entry_archive.m_context` might be None, and all the functionalities in ELNJupyterAnalysis depend on the m_context.

**This is exactly what I meant in my first comment to Hampus**, thus, and in order to stop losing time commenting on each other, this test should be skipped (for now).

Btw, everything you just replied, @ka-sarthak , implies touching functionalities from the core nomad package, which is what this testing should not be about.

ka-sarthak replied:
The points you are raising are quite valid. It seems that when `nomad.client.normalize_all` is run over the schema, the entry archive which is generated does not have all the metadata: `archive.metadata` only contains `mainfile` property, all the other properties that are supposed to be generated by nomad are missing. In that sense, performing an "integration testing" that runs all the normalize functions is not supported nicely.

Regarding the logger (point 1), that can be bypassed using a pytest fixture that goes over the log data to find errors.

While `nomad.client.normalize_all` already goes through all associated normalizers, I think we can also make unit tests for each of the normalize function (to divide the test into smaller components). As I understand, this is also your opinion @JosePizarro3. For example, having a test `test_ELNJupyterNotebookAnalysis` and testing it's `normalize` function. But the issue with missing metadata will still bite us if the normalization of this class or one of its parents uses something from metadata (for example `name` quantity of class `BaseSection`). 
@hampusnasstrom what do you think?

hampusnasstrom commented on MANIFEST.in:
```diff
@@ -1 +0,0 @@
-recursive-include * nomad_plugin.yaml
```
This should be included again.

hampusnasstrom commented on README.md:
```diff
@@ -1,77 +1,85 @@
-![](https://github.com/nomad-coe/nomad-schema-plugin-example/actions/workflows/actions.yml/badge.svg)
-![](https://coveralls.io/repos/github/FAIRmat-NFDI/nomad-analysis/badge.svg?branch=main)
```
What is the reason for removing the coveralls badge?

JosePizarro3 replied:
My bad, I will include it back again (same for the badge of the workflow)

hampusnasstrom commented on README.md:
```diff
@@ -1,77 +1,85 @@
-![](https://github.com/nomad-coe/nomad-schema-plugin-example/actions/workflows/actions.yml/badge.svg)
```
This should be switched to the workflow of this repo but would be good to include.

JosePizarro3 replied:
See above.

hampusnasstrom commented on pyproject.toml:
```diff
@@ -1,65 +1,57 @@
 
 [project]
 name = "nomad-analysis"
-version = '0.1.0'
```
I would setup dynamic versioning here like we have in `nomad-material-processing` and `nomad-measurements` packages.

JosePizarro3 replied:
Yeah, agree, I will change it to `dynamic = ["version"]`.

hampusnasstrom replied:
You need to also add the `setuptools-scm` config for this. You can see how I did it here for example: [https://github.com/FAIRmat-NFDI/nomad-material-processing/blob/main/pyproject.toml](https://github.com/FAIRmat-NFDI/nomad-material-processing/blob/main/pyproject.toml)

hampusnasstrom reviewed (changes requested):
I was listed under "reviewers" so I guess that meant you wanted me to review this @ka-sarthak? Added some comments and suggestions.

JosePizarro3 committed:
Restore MANIFEST.in file

Restore README development section

Added setuptools-scm and Documentation url to pyproject

JosePizarro3 committed:
Restored plural in yaml name

JosePizarro3 committed:
Fix pyproject and pipeline

ka-sarthak commented on src/nomad_analysis/nomad_plugin.yaml:
```diff
@@ -1,4 +1,4 @@
-  This plugin is used to analyze parsed data files.
\ No newline at end of file
+description: |
+  This plugin is used to analyze parsed raw data for spectral profiles in the standard NOMAD schema.
```
Should we not keep the description more general? As I understand, spectral profiles will one part (perhaps a big and important one) of the whole idea of analysis

JosePizarro3 replied:
Agree. I changed it to:

  This plugin is used to analyze parsed raw data from the NOMAD standard schema.

ka-sarthak commented on README.md:
````diff
@@ -1,77 +1,95 @@
+name: schemas/nomad_analysis
+description: |
+  This plugin is used to analyze parsed raw data for spectral profiles in the standard NOMAD schema.
+```
````
As this content is already in `src/nomad_analysis/nomad_plugin.yaml`, do we need this in README?

JosePizarro3 replied:
Well, being a fork from the parser example, I would keep it. But yes, I admit it is too much info here.

JosePizarro3 replied:
Ok, check it out. I think it is more clear that the content is already there.

ka-sarthak replied:
Looks good to me! 
Please make the description same as the one in the `nomad_plugin.yaml` file.
> This plugin is used to analyze parsed raw data from the NOMAD standard schema.

JosePizarro3 committed:
Test pipeline with string

JosePizarro3 committed:
Added last comments

Restore test_schema.py

JosePizarro3 committed:
Deleted launch.json after rebasing

JosePizarro3 committed:
Apply ruff formatting

JosePizarro3 commented:
@ka-sarthak this is ready I think.

Please, note the Ruff formatting for your branches in the future 🙂 The pipeline now works as it should regarding `ruff format . --check`.

ka-sarthak commented on tests/test_schema.py:
```diff
@@ -1,3 +1,21 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import pytest
```
You can also remove the `pytest` import as we are not using it

ka-sarthak commented on README.md:
```diff
@@ -1,77 +1,95 @@
 
 # NOMAD's Analysis plugin
-This is a plugin for [NOMAD](https://nomad-lab.eu) to facilitate analysis of processed entry archives using Jupyter notebooks.
+This is a plugin for [NOMAD](https://nomad-lab.eu) to facilitate analysis of processed entry archives using classes and functions which can then be import for using in Jupyter notebooks.
```
replace ~import~ with _imported_ 
Also, consider this re-writing for more clarity:
> This is a plugin for [NOMAD](https://nomad-lab.eu) to facilitate analysis of processed entry archives using classes and functions. The plugin also supports ELN schemas which allow to use these functions in Jupyter notebooks, and persist the analysis workflow.

ka-sarthak commented on README.md:
```diff
@@ -1,77 +1,95 @@
 ### Run the tests
 
-You can run automated tests with `pytest`:
+You can the unit testing using the `pytest` package:
```
I think there's a missing word:
> You can *run* the unit testing ...

ka-sarthak reviewed (approved):
Looks great to me!! :smile: 

I have added some comments on the README containing some minor fixes and rewriting suggestion. Also, there's an import statement in the test_schema.py that can be removed.

I am approving it, but please make sure to integrate the review before merging.

I also think you will have to rebase to the current develop before merging.

JosePizarro3 committed:
Final cleanup

JosePizarro3 committed:
Fix after rebasing

JosePizarro3 committed:
Fix ruff

JosePizarro3 committed:
I hate when git gets crazy after rebasing.... -.-

JosePizarro3 committed:
Fixed dockerfile
